File: .portkit/scripts/typescript/sanitize-upstream.ts
import { Project } from "ts-morph";
import * as path from "path";
import * as fs from "fs";

// Simple "Poison" patterns to remove/rewrite
// Unlike Python regex, we can potentially use AST here to be safer.
// But for "Sanitization", often removing lines is valid.

function sanitizeUpstream(targetFile: string) {
    if (!fs.existsSync(targetFile)) {
        console.error("File not found");
        process.exit(1);
    }

    // Load Rules
    const rulesPath = path.resolve(".portkit/ecosystem-rules.json");
    let deniedDeps: any[] = [];
    if (fs.existsSync(rulesPath)) {
        try {
            const rules = JSON.parse(fs.readFileSync(rulesPath, "utf-8"));
            deniedDeps = rules.denied_dependencies || [];
        } catch (e) {
            console.warn("Failed to load ecosystem rules", e);
        }
    }

    let content = fs.readFileSync(targetFile, "utf-8");

    // Dynamic Sanitization
    for (const dep of deniedDeps) {
        if (dep.action === "strip") {
            const pkg = dep.package;
            // import ... from "pkg"
            const regexImport = new RegExp(`import\\s+.*?from\\s+['"]${pkg}['"];?`, "gm");
            content = content.replace(regexImport, "");

            // import "pkg"
            const regexSideEffect = new RegExp(`import\\s+['"]${pkg}['"];?`, "gm");
            content = content.replace(regexSideEffect, "");
        }
    }

    // Write back
    fs.writeFileSync(targetFile, content, "utf-8");
    console.log(`Sanitized: ${targetFile}`);
}

const args = process.argv.slice(2);
const target = args[0]; // First arg is file

if (target) {
    sanitizeUpstream(target);
} else {
    console.error("Usage: ts-node sanitize-upstream.ts [FILE]");
    process.exit(1);
}

File: .portkit/scripts/python/sanitize-upstream.py
import argparse
import re
import sys
import json
from pathlib import Path

def load_patterns():
    patterns = []
    rules_path = Path('.portkit/ecosystem-rules.json')
    if rules_path.exists():
        try:
            with open(rules_path, 'r', encoding='utf-8') as f:
                data = json.load(f)
                for dep in data.get('denied_dependencies', []):
                    pkg = dep.get('package')
                    action = dep.get('action')
                    
                    if action == 'strip':
                        # Generalized Import Removal
                        # import ... from "pkg"
                        # import "pkg"
                        pat = rf"import\s+.*?from\s+['\"]{pkg}['\"];?"
                        patterns.append((pat, ""))
                        pat_side_effect = rf"import\s+['\"]{pkg}['\"];?"
                        patterns.append((pat_side_effect, ""))
                        
        except Exception as e:
            print(f"Warning: Failed to load ecosystem rules: {e}")
    return patterns

def sanitize(content):
    modified = content
    patterns = load_patterns()
    
    for pattern, replacement in patterns:
        modified = re.sub(pattern, replacement, modified, flags=re.MULTILINE)
    
    return modified

def main():
    parser = argparse.ArgumentParser(description="Sanitize upstream code.")
    parser.add_argument("file", help="File to sanitize (edits in place)")
    args = parser.parse_args()

    target = Path(args.file)
    if not target.exists():
        print("File not found.")
        sys.exit(1)

    with open(target, 'r', encoding='utf-8') as f:
        content = f.read()

    new_content = sanitize(content)

    if new_content != content:
        with open(target, 'w', encoding='utf-8') as f:
            f.write(new_content)
        print(f"Sanitized: {target}")
    else:
        print(f"Clean: {target}")

if __name__ == "__main__":
    main()
